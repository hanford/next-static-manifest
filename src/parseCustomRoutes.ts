import findUp from 'find-up';
import { CONFIG_FILE } from 'next/constants';
import loadCustomRoutes, {
  normalizeRouteRegex,
} from 'next/dist/lib/load-custom-routes';
import { defaultConfig } from 'next/dist/server/config-shared';
import { pathToRegexp } from 'path-to-regexp';

// import { RouteEntry } from './utils';

function createMap(route: { source: string; destination: string }) {
  const routeRegex = pathToRegexp(route.source, [], {
    strict: true,
    sensitive: false,
    delimiter: '/', // default is `/#?`, but Next does not pass query info
  });

  let regexSource = routeRegex.source;

  return {
    src: route.source,
    dest: route.destination,
    regex: normalizeRouteRegex(regexSource),
    dynamic: true,
  };
}

export default async function parseCustomRoutes(dir: string) {
  const path = await findUp(CONFIG_FILE, { cwd: dir });

  if (path?.length) {
    const userConfigModule = require(path);

    const userConfig = { ...defaultConfig, ...userConfigModule };

    const { rewrites: _rewrites } = await loadCustomRoutes(userConfig);

    const rewrites = {
      beforeFiles: _rewrites.beforeFiles.map(createMap),
      afterFiles: _rewrites.afterFiles.map(createMap),
    };

    return rewrites;
  }

  return { beforeFiles: [], afterFiles: [] };
}
