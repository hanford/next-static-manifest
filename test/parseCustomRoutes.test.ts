import parseCustomRoutes from '../src/parseCustomRoutes';

describe('reads next.config.js and respects rewrites fn', () => {
  it('generates RouteEntry from next.config.rewrites', async () => {
    const routes = await parseCustomRoutes(__dirname);

    expect(routes).toMatchSnapshot();
  });
});
