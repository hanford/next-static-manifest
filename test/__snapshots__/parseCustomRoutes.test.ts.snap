// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reads next.config.js and respects rewrites fn generates RouteEntry from next.config.rewrites 1`] = `
Object {
  "afterFiles": Array [
    Object {
      "dest": "http://foo.com/playground",
      "dynamic": true,
      "regex": "^/playground$",
      "src": "/playground",
    },
    Object {
      "dest": "http://foo.com/playground/:path*",
      "dynamic": true,
      "regex": "^/playground(?:/((?:[^/]+?)(?:/(?:[^/]+?))*))?$",
      "src": "/playground/:path*",
    },
  ],
  "beforeFiles": Array [],
}
`;

exports[`reads next.config.js and respects rewrites fn works with async/await 1`] = `
Object {
  "afterFiles": Array [
    Object {
      "dest": "http://foo.com/playground",
      "dynamic": true,
      "regex": "^/playground$",
      "src": "/playground",
    },
    Object {
      "dest": "http://foo.com/playground/:path*",
      "dynamic": true,
      "regex": "^/playground(?:/((?:[^/]+?)(?:/(?:[^/]+?))*))?$",
      "src": "/playground/:path*",
    },
  ],
  "beforeFiles": Array [],
}
`;
